### ISSUE TITLE
Support multiple error types in the Python and JS clients

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#4472]
https://github.com/rethinkdb/rethinkdb/issues/4296 implemented server support, but we have to actually produce multiple exception types on the clients.  (I already did this for Ruby.)

**Comment by danielmewes** [comment]
Moving to 2.1. I think we should support this across all official drivers.

**Comment by marshall007** [comment]
I have a pretty good start on this for the JS driver in [`marshall007/rethinkdb#marshall_4472_js`](https://github.com/marshall007/rethinkdb/tree/marshall_4472_js). Haven't tested yet since they need to be run against 2.1.x server.

**Comment by marshall007** [comment]
I wasn't able to build the server binary off of `next` to test against:

```
    ...
    [317/397] CC build/release_clang/obj/rdb_protocol/query_server.o
src/rdb_protocol/query_server.cc:56:65: error: too few arguments to function call, expected 5, have 4
                       ql::backtrace_registry_t::EMPTY_BACKTRACE);
                                                                ^
./src/rdb_protocol/backtrace.hpp:72:1: note: 'fill_error' declared here
void fill_error(Response *res_out,
^
1 error generated.
make[1]: *** [build/release_clang/obj/rdb_protocol/query_server.o] Error 1
make: *** [make] Error 2
```

Would it be easier if I just did this directly on top of @mlucy's changes (i.e. 1513a25)?

**Comment by mlucy** [comment]
@marshall007 -- it would probably be slightly better to do it off of `next`.  It looks like I introduced the bug myself in the error types merge; just fixed it and pushed, so `next` should compile fine in release mode now.

**Comment by marshall007** [comment]
Ok, thanks. I'll rebase and give it another go.

**Comment by marshall007** [comment]
@mlucy I didn't come across any `RqlOpIndeterminateError` exceptions and want to make sure that wasn't me. Take a look at my comment over there when you get a chance.

**Comment by neumino** [comment]
Are the new errors just sub-types of a `ReqlRuntimeError`?
Can we have a quick description of each type? #4296 doesn't say much about what was implemented

**Comment by marshall007** [comment]
@neumino pretty much but there's a couple cases of multiple inheritance. Take a look at the [relevant lines](https://github.com/marshall007/rethinkdb/blob/marshall_4472_js/drivers/javascript/errors.coffee#L24-L38) from my PR.

**Comment by danielmewes** [comment]
Thanks to @marshall007 , we only have the Python driver remaining now.

**Comment by VeXocide** [comment]
In CR 3080 by @larkost.

**Comment by VeXocide** [comment]
Merged into `next` via commit ecf418d26f2beae1243fe3e35db834bbb5e120cf.
